python/src/equistore/operations/join.py

```diff
@@ -0,0 +1,236 @@
+        else:
+            properties1 = np.vstack(
+                [np.zeros(len(properties1), dtype=int), np.arange(len(properties1))]
+            ).T
+
+        if "tensor" == names2[0]:
+            # if there is already also `tensor` property we keep it and
+            # increase the values of the second block by the max value of
+            # the first
+            properties2 = np.vstack(
+                [
+                    properties2[:, 0] + 1 + properties1[:, 0].max(),
+                    np.arange(len(properties2)),
+                ]
+            ).T
+        else:
+            properties2 = np.vstack(
+                [
+                    np.zeros(len(properties2), dtype=int) + 1 + properties1[:, 0].max(),
+                    np.arange(len(properties2)).T,
+                ]
+            ).T
+
+        property_names = ("tensor", "property")
+    elif names1 == names2 and len(np.intersect1d(labels_1, labels_2)) > 0:
+        # if property_names in each block are the same but the values are not
+        # unique we keep the names and add an extra dimension for each tensor
+
+        if "tensor" == names1[0]:
+            # if there is already a `tensor` property we increase the values
+            # of the second block by the max value of the first
+            property_names = names1
+            properties2[:, 0] += 1 + properties1[:, 0].max()
+        else:
+            property_names = ("tensor",) + names1
+            properties1 = np.hstack(
+                [np.zeros(len(properties1), dtype=int).reshape(-1, 1), properties1]
+            )
+            properties2 = np.hstack(
+                [np.ones(len(properties2), dtype=int).reshape(-1, 1), properties2]
+            )
+    else:
+        # if property_names are the same in each block and the values are
+        # unique we keep the names of the first block and stack the values
+        property_names = names1
+
+    properties_values = np.vstack([properties1, properties2])
+
+    return Labels(names=property_names, values=properties_values)
+
+
+def _join_blocks_along_properties(
+    block1: TensorBlock, block2: TensorBlock
+) -> TensorBlock:
+    if not np.all(block1.samples == block2.samples):
+        raise ValueError("`A` and `B` do not have the same samples.")
+
+    if block1.components != block2.components:
+        raise ValueError("`A` and `B` do not have the same components")
+
+    if sorted(block1.gradients_list()) != sorted(block2.gradients_list()):
+        raise ValueError("`A` and `B` do not have the same gradients")
+
+    properties = _join_labels(block1.properties, block2.properties)
+    result_block = TensorBlock(
+        values=_dispatch.hstack([block1.values, block2.values]),
+        samples=block2.samples,
+        components=[],
+        properties=properties,
+    )
+
+    for parameter, gradient1 in block1.gradients():
+        gradient2 = block2.gradient(parameter)
+
```

Comment:
this should also check that the gradients components are the same

Reply:
Yes!